[AutoDiff] Fix `@differentiating` type-checking for differentials.

dan-zheng opened the pull request:
Fix type-checking for differential functions.
Simplify pullback/differential type-checking diagnostics.

Resolves [TF-366](https://bugs.swift.org/browse/TF-366).

dan-zheng committed:
[AutoDiff] Fix `@differentiating` type-checking for differentials.

Fix type-checking for differential functions.
Simplify pullback/differential type-checking diagnostics.

Resolves TF-366.

dan-zheng commented:
@swift-ci Please test tensorflow

rxwei commented on test/AutoDiff/differentiating_attr_type_checking.swift:
```diff
@@ -110,11 +116,12 @@ extension AdditiveArithmetic where Self : Differentiable, Self == Self.Cotangent
 extension InstanceMethod {
   // If `Self` conforms to `Differentiable`, then `Self` is currently always inferred to be a differentiation parameter.
-  // expected-error @+1 {{unexpected differentiation parameters type; got 'Self.CotangentVector', but expected '(Self.CotangentVector, Self.CotangentVector)'}}
+  // expected-error @+1 {{unexpected 'pullback' type; got '(Self.CotangentVector) -> Self.CotangentVector', but expected '(Self.CotangentVector) -> (Self.CotangentVector, Self.CotangentVector)'}}
```
Ideally, there should also be a note emitted at the pullback type signature.

dan-zheng replied:
How about changing the error to show up on the pullback/differential type signature?
```
tf-366-3.swift:9:70: error: unexpected 'pullback' type; got '(Self.CotangentVector) -> T.CotangentVector', but expected '(Self.CotangentVector) -> (Self.CotangentVector, T.CotangentVector)'
  func vjpFoo<T : Differentiable>(_ x: T) -> (value: Self, pullback: (Self.CotangentVector) -> T.CotangentVector) {
                                                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

The error doesn't clearly mention `@differentiating` attribute though.

dan-zheng replied:
Updated in https://github.com/apple/swift/pull/23353/commits/d08e50d8dd137a97fd5bacb2c11ea2c762db0970.

Examples:
```
tf-366-3.swift:9:70: error: 'pullback' does not have expected type '(Self.CotangentVector) -> (Self.CotangentVector, T.CotangentVector)'
  func vjpFoo<T : Differentiable>(_ x: T) -> (value: Self, pullback: (Self.CotangentVector) -> T.CotangentVector) {
                                                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
tf-366-3.swift:19:74: error: 'differential' does not have expected type '(Self.TangentVector, T.TangentVector) -> Self.TangentVector'
  func jvpFoo<T : Differentiable>(_ x: T) -> (value: Self, differential: (Self.TangentVector, T.TangentVector) -> T.TangentVector) {
                                                                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

rxwei replied:
The proposed solution does not follow the “whoever caused the error should get an error” rule, which is being followed by the differentiation transform and other parts of the compiler. The function declaration is not wrong — it’s valid Swift code. It’s the @differentiating attribute that caused the function to not compile, and thus the attribute should get the error.

dan-zheng replied:
Gotcha. How about this (any diagnostic message suggestions)?
```
tf-366-3.swift:8:4: error: unexpected 'pullback' type
  @differentiating(foo)
   ^
tf-366-3.swift:9:70: note: 'pullback' does not have expected type '(Self.CotangentVector) -> (Self.CotangentVector, T.CotangentVector)'
  func vjpFoo<T : Differentiable>(_ x: T) -> (value: Self, pullback: (Self.CotangentVector) -> T.CotangentVector) {
                                                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
tf-366-3.swift:18:4: error: unexpected 'differential' type
  @differentiating(foo)
   ^
tf-366-3.swift:19:74: note: 'differential' does not have expected type '(Self.TangentVector, T.TangentVector) -> Self.TangentVector'
  func jvpFoo<T : Differentiable>(_ x: T) -> (value: Self, differential: (Self.TangentVector, T.TangentVector) -> T.TangentVector) {
                                                                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

rxwei replied:
* It would also be good to cross reference the original function, because when `foo` is mistyped (or overloaded), it is unclear which `foo` it is referring to. Could you show another note at the original function's definition (if its location is valid and is in the same file)?
* I suggest changing the error to "function's 'pullback' result does not match `foo` (a.k.a. `foo(_:_:)`)" (show the fully qualified decl name in "aka").

dan-zheng replied:
> * It would also be good to cross reference the original function, because when `foo` is mistyped (or overloaded), it is unclear which `foo` it is referring to. Could you show another note at the original function's definition (if its location is valid and is in the same file)?

What message do you have in mind for the note at the original function?

> * I suggest changing the error to "function's 'pullback' result does not match `foo` (a.k.a. `foo(_:_:)`)" (show the fully qualified decl name in "aka").

I'll make this change.

> * Can you move the error to the location of `@` (I think it's called `AtLoc` or something), and highlight the entire attribute?

Yes, this is possible and looks nicer, though unprecedented - all existing diagnostics use `attr->getLocation` which shows up in the current ugly spot. I'll go ahead and use the `@` loc and highlight the attribute.

rxwei replied:
something like "`foo(_:_:)` defined here"

dan-zheng replied:
> Something like "`foo(_:_:)` defined here" is fine.

Great, I'll implement this. 

> If unprecedented, we should ask the forum about it!

Sounds good! I'll hold off on changing the error location then and keep `attr->getLocation`.

dan-zheng replied:
Changes made in https://github.com/apple/swift/pull/23353/commits/a8c946d4c2deaa9877beb728aff4986110d79df6:

```
- Revamp differential/pullback type mismatch diagnostics.
  - Emit type mismatch error on `@differentiating` attribute.
  - Emit note with expected differential/pullback type on actual type location.
  - Emit note showing original function location, if possible.
- Perform original function lookup before checking `wrt` parameters.
  - Original function lookup should occur as early as possible because
    "original function not found" errors should be emitted earlier than other
    errors.
- Change error message for `differentiating_attr_overload_not_found`.
  - The blame for "original function not found" should be placed on the
    `@differentiating` derivative function, not the original function.
  - Old: "'original' does not have expected type ..."
  - New: "could not find 'original' with expected type ..."
```

I didn't show fully qualified names for original functions because that's unprecedented: all `DeclName`s for normal functions are printed without argument names:
```cpp
// Argument names are printed only for "special" DeclNames, which are
// subscripts/constructors/destructors.
llvm::raw_ostream &DeclName::printPretty(llvm::raw_ostream &os) const {
  return print(os, /*skipEmptyArgumentNames=*/!isSpecial());
}
```

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng committed:
Change diagnostic to highlight pullback/differential type signature.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng committed:
Revamp differential/pullback type mismatch diagnostics.

- Revamp differential/pullback type mismatch diagnostics.
  - Emit type mismatch error on `@differentiating` attribute.
  - Emit note with expected differential/pullback type on actual type location.
  - Emit note showing original function location, if possible.
- Perform original function lookup before checking `wrt` parameters.
  - Original function lookup should occur as early as possible because
    "original function not found" errors should be emitted earlier than other
    errors.
- Change error message for `differentiating_attr_overload_not_found`.
  - The blame for "original function not found" should be placed on the
    `@differentiating` derivative function, not the original function.
  - Old: "'original' does not have expected type ..."
  - New: "could not find 'original' with expected type ..."

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow
